In [0]:
dataPath = "/FileStore/tables/df_silver.csv"
df0 = spark.read.format("csv")\
  .option("header","true")\
  .option("inferSchema", "true")\
  .load(dataPath)

display(df0)

name review_count categories rating latitude price distance longitude postal_code El Sur 769 Spanish 4.5 40.41105 2 1323.9351440693945 -3.6995454 28012 Alhambra 448 Spanish 4.0 40.41587 2 847.3790987480967 -3.7016 28012 La Bodega de Los Secretos 31 Spanish 4.5 40.410683 3 1375.936331500051 -3.694688 28014 Casa Julio 42 Spanish 4.5 40.42427 2 546.9655482728417 -3.70368 28004 Celso y Manolo 182 Spanish 4.5 40.420185 2 297.6040361411656 -3.6974769 28004 Taberna El Sur de Huertas 41 Spanish 5.0 40.41366 null 1036.183034652163 -3.69943 28014 Malacatín 24 Spanish 4.5 40.41031 null 1638.679922735727 -3.70765 28005 Más Al Sur 126 Spanish 4.5 40.41002 2 1409.3953170255295 -3.69693 28012 La Casa del Abuelo 161 Spanish 4.0 40.415768 2 859.0034856078098 -3.701606 28012 Oink 16 Spanish 4.5 40.42004 1 398.07076724778057 -3.700481 28013 Street Xo 94 Spanish 4.5 40.427647 3 1035.445088896326 -3.687092 28001 Vinoteca Moratín 33 Spanish 5.0 40.41252 3 1168.8084348720095 -3.69541 28014 El Rincón de Ati 8 Spanish 4.5 40.415123 2 1257.8650825827144 -3.708427 28005 La Mi Venta 102 Spanish 4.5 40.420456 3 1282.4301695469583 -3.7124026 28013 Bodega de la Ardosa 125 Spanish 4.0 40.42375 2 371.17219437757063 -3.70182 28004 Algarabía 16 Spanish 5.0 40.41732 2 1245.5718114622448 -3.7103744 28013 Lateral 79 Spanish 4.0 40.42306 2 299.75414566377356 -3.70113 28004 Entrevinos 11 Spanish 4.5 40.42615 2 1679.1324348217292 -3.71713 28008 Los Montes de Galicia 54 Spanish 4.5 40.4346 3 2810.2129913861268 -3.66854 28028 Sacha 12 Spanish 4.5 40.460682 4 4292.8105582067265 -3.6874075 28036 Malaspina 210 Spanish 4.0 40.41582 1 880.0271748434934 -3.70253 28012 Tierra del Queiles 20 Spanish 4.5 40.42661 2 888.0611582031488 -3.68828 28001 Casa Toni 86 Spanish 4.5 40.415554 1 889.604983373482 -3.7018628 28012 La Austriaca 23 Spanish 4.5 40.42139 2 371.7496282903476 -3.70149 28004 Ciudad de Tui 26 Spanish 4.0 40.41823 null 659.7646610690329 -3.70241 28013 O Pazo de Lugo 55 Spanish 4.0 40.407417 2 1720.5167688397755 -3.6962805 28012 El Comunista 9 Spanish 4.5 40.42219 2 70.33818378885029 -3.69756 28004 La Garriga 2 Spanish 5.0 40.426544 null 560.7903673495854 -3.69306 28004 Selva 9 Spanish 4.5 40.423363 2 936.2070628085398 -3.708626 28015 Rosi La Loca 180 Spanish 4.5 40.415813 3 906.7764467348135 -3.7029788 28012 El Lagar 14 Spanish 4.5 40.42866 1 2003.3739086320672 -3.71991 28008 La Descubierta 73 Spanish 4.0 40.41525 1 946.019092804493 -3.70262 28012 Casa Amadeo 15 Spanish 4.5 40.41052 1 1606.8643189404181 -3.70749 28005 La Gastro 69 Spanish 4.5 40.42129 2 488.7083814997758 -3.70289 28004 El Tigre 366 Spanish 3.5 40.420254 1 290.9915674469935 -3.6979134 28004 La Española 8 Spanish 4.5 40.420994 null 320.6234167699696 -3.700475 28004 El Asador de Aranda 29 Spanish 4.5 40.41985 3 922.8203902039091 -3.70764 28013 Botín 615 Spanish 4.0 40.41406 3 1304.2170679660362 -3.70803 28005 El Barril 21 Spanish 5.0 40.42199 3 630.0871640719441 -3.69023 28001 La Perejila 32 Spanish 4.5 40.41229 2 1530.886543775833 -3.70924 28005 Tribuetxe 20 Spanish 4.5 40.407425 1 1780.9484656935279 -3.703198 28012 Los Arcos 6 Spanish 5.0 40.43787 3 1675.278430632665 -3.6992831 28010 La Dominga 26 Spanish 4.0 40.42526 2 567.3408517478088 -3.70348 28004 La Tasqueria 21 Spanish 5.0 40.42228 3 2019.6198202824544 -3.67374 28009 Angelita 29 Spanish 5.0 40.420277 3 372.5374556274898 -3.700387 28004 Casa Dani 35 Spanish 4.5 40.426846 1 1094.0981652402868 -3.6857686 28001 La Sanabresa 52 Spanish 4.5 40.41254 2 1148.4834173354568 -3.69854 28014 Casa Lucas 154 Spanish 4.5 40.412327 2 1541.9081259234936 -3.709429 28005 El Cogollo 20 Spanish 4.5 40.417 2 1106.3064051275262 -3.7081468 28013 Lakasa 48 Spanish 5.0 40.440876 2 2020.5902312080937 -3.700715 28003 Steak Burger 3 Burgers 4.5 40.41983 null 885.8031586140268 -3.707263 28013 Bacoa 67 Burgers 4.5 40.4167 2 764.7506754570883 -3.701595 28014 Viva Burger 34 Burgers 4.5 40.41247 2 1683.8474441060912 -3.71181 28005 Steak Burger 21 Burgers 4.0 40.419746 null 379.9

In [0]:
from pyspark.sql.functions import col,length

#Here we notice that there are some null values for the "price" column: there are some restaurants that don't have a price feature (so how expensive or cheap they are). Let's see how many are there to decide how to clean them (remove or mean imputation or regression...)
df0.where(length(col("price")) > 1).count()


Out[2]: 45

We notice that there are 45 values which are missing for price. Since price is an important variable as we saw in the EDA, we will not remove it. Plus, we have very few variables, so removing one would not be rational. Also, imputing would bias the data in our opinion, so we will just proceed to remove the rows with these nulls

In [0]:
#Remove rows with nulls
df0 = df0.where(length(col("price")) < 2)

#Also remove columns that we won't use: name (no use), postal code (no use), longitude and latitude (because we have distance)
df0 = df0.drop('name')
df0 = df0.drop('postal_code')
df0 = df0.drop('longitude')
df0 = df0.drop('latitude')
df0.write.csv('Newrestsss.csv', header=True)

In [0]:
# Load the CSV file from the data section of Databricks
df = spark.read.csv('/Newrestsss.csv', header=True)

In [0]:
from pyspark.ml.feature import StringIndexer
# We only have 1 categorical variable, which is "categories", so we just labelEncode it using StringIndexer
indexer = StringIndexer(inputCol='categories', outputCol='categories_index')
df = indexer.fit(df).transform(df)

In [0]:
#We have to make sure that all of the variables are numerical for the ML process
df = df.withColumn("rating", df["rating"].cast('double'))
df = df.withColumn("review_count", df["review_count"].cast('double'))
# df = df.withColumn("latitude", df["latitude"].cast('double'))
df = df.withColumn("price", df["price"].cast('double'))
df = df.withColumn("distance", df["distance"].cast('double'))
# df = df.withColumn("postal_code", df["postal_code"].cast('double'))
df = df.withColumn("rating", df["rating"].cast('double'))
# df = df.withColumn("longitude", df["longitude"].cast('double'))
df = df.withColumn("categories_index", df["categories_index"].cast('double'))
df = df.drop('categories')
display(df)

review_count rating price distance categories_index 769.0 4.5 2.0 1323.9351440693945 0.0 448.0 4.0 2.0 847.3790987480967 0.0 31.0 4.5 3.0 1375.936331500051 0.0 42.0 4.5 2.0 546.9655482728417 0.0 182.0 4.5 2.0 297.6040361411656 0.0 126.0 4.5 2.0 1409.3953170255295 0.0 161.0 4.0 2.0 859.0034856078098 0.0 16.0 4.5 1.0 398.07076724778057 0.0 94.0 4.5 3.0 1035.445088896326 0.0 33.0 5.0 3.0 1168.8084348720095 0.0 8.0 4.5 2.0 1257.8650825827144 0.0 102.0 4.5 3.0 1282.4301695469583 0.0 125.0 4.0 2.0 371.17219437757063 0.0 16.0 5.0 2.0 1245.5718114622448 0.0 79.0 4.0 2.0 299.75414566377356 0.0 11.0 4.5 2.0 1679.1324348217292 0.0 54.0 4.5 3.0 2810.2129913861268 0.0 12.0 4.5 4.0 4292.8105582067265 0.0 210.0 4.0 1.0 880.0271748434934 0.0 20.0 4.5 2.0 888.0611582031488 0.0 86.0 4.5 1.0 889.604983373482 0.0 23.0 4.5 2.0 371.7496282903476 0.0 55.0 4.0 2.0 1720.5167688397755 0.0 9.0 4.5 2.0 70.33818378885029 0.0 9.0 4.5 2.0 936.2070628085398 0.0 180.0 4.5 3.0 906.7764467348135 0.0 14.0 4.5 1.0 2003.3739086320672 0.0 73.0 4.0 1.0 946.019092804493 0.0 15.0 4.5 1.0 1606.8643189404181 0.0 69.0 4.5 2.0 488.7083814997758 0.0 366.0 3.5 1.0 290.9915674469935 0.0 29.0 4.5 3.0 922.8203902039091 0.0 615.0 4.0 3.0 1304.2170679660362 0.0 21.0 5.0 3.0 630.0871640719441 0.0 32.0 4.5 2.0 1530.886543775833 0.0 20.0 4.5 1.0 1780.9484656935279 0.0 6.0 5.0 3.0 1675.278430632665 0.0 26.0 4.0 2.0 567.3408517478088 0.0 21.0 5.0 3.0 2019.6198202824544 0.0 29.0 5.0 3.0 372.5374556274898 0.0 35.0 4.5 1.0 1094.0981652402868 0.0 52.0 4.5 2.0 1148.4834173354568 0.0 154.0 4.5 2.0 1541.9081259234936 0.0 20.0 4.5 2.0 1106.3064051275262 0.0 48.0 5.0 2.0 2020.5902312080937 0.0 67.0 4.5 2.0 764.7506754570883 3.0 34.0 4.5 2.0 1683.8474441060912 3.0 28.0 4.5 2.0 1185.7862478138004 3.0 180.0 4.0 2.0 1090.3118768624156 3.0 25.0 4.0 2.0 1506.045945872656 3.0 54.0 4.0 2.0 609.4105630939849 3.0 25.0 4.0 1.0 1022.4985490488237 3.0 17.0 4.0 2.0 724.1376317449414 3.0 51.0 4.5 2.0 1523.7896540257736 3.0 33.0 3.5 2.0 741.9874349730312 3.0 24.0 4.5 1.0 3075.7635209239843 3.0 45.0 4.0 2.0 1984.3728826077909 3.0 25.0 4.0 1.0 492.6528185606649 3.0 70.0 4.5 1.0 1886.4628184569287 3.0 4.0 4.5 2.0 5625.145609505655 3.0 18.0 4.0 2.0 857.9076724337576 3.0 5.0 4.0 2.0 95.65511145937545 3.0 11.0 4.5 2.0 440.75710302522737 3.0 109.0 4.0 2.0 772.9589767071259 3.0 9.0 4.5 2.0 4199.438176581559 3.0 12.0 4.0 2.0 835.9441616287093 3.0 10.0 3.5 1.0 1977.934394894583 3.0 41.0 4.0 2.0 840.091922029471 3.0 73.0 4.0 2.0 181.88790341618449 3.0 7.0 4.0 2.0 1904.7813667856042 3.0 7.0 4.0 2.0 327.0667231780543 3.0 21.0 4.0 1.0 1076.5736060546099 3.0 153.0 4.5 2.0 907.9863770759232 3.0 20.0 4.5 2.0 1143.2003384440857 3.0 33.0 4.5 2.0 156.68651580601073 3.0 7.0 4.0 2.0 5812.567773122338 3.0 60.0 4.5 1.0 947.1811780352955 3.0 50.0 4.0 2.0 572.505669037329 3.0 2.0 4.0 2.0 6016.154755508205 3.0 74.0 4.0 2.0 1124.0231004567572 1.0 111.0 4.5 2.0 127.1759631338847 1.0 31.0 4.5 3.0 213.80119474053967 1.0 44.0 4.5 2.0 680.121843136792 1.0 51.0 4.0 2.0 1066.865619950464 1.0 15.0 4.5 2.0 750.5585941275872 1.0 18.0 4.5 2.0 977.7756847131163 1.0 46.0 4.5 1.0 336.6517267620623 1.0 6.0 4.5 3.0 467.4383786947757 1.0 51.0 4.5 2.0 366.44865696195694 1.0 38.0 4.5 2.0 939.3995899152179 1.0 30.0 4.0 2.0 391.24137975900635 1.0 49.0 4.5 1.0 1135.4881319340707 1.0 4.0 5.0 3.0 1259.3239562337224 1.0 20.0 4.5 1.0 1590.6244316723469 1.0 22.0 4.5 2.0 2133.383536508534 1.0 7.0 4.5 2.0 2077.6147433546594 1.0 12.0 3.5 3.0 1132.2225436036554 1.0 16.0 4.0 2.0 1456.4016545306733 1.0 4.0 4.5 2.0 1251.9185393748037 1.0 2.0 4.5 3.0 106.04678146607492 1.0 29.0 5.0 3.0 372.5374556274898 1.0 12.0 4.0 2.0 352.8375009931065 1.0 4.0 4.0 2.0 609.1326001777951 1.0 4.0 5.0 1.0 1095.2377715771497 1.0 1.0 5.0 1.0 989.864446124471 1.0 6.0 4.0 3.0 393.98918911987215 1.0 13.0 4.0 2.0 1286.5170897886294 1.0 24.0 4.0 1.0 2228.9328208031056 1.0 3.0 4.5 1.0 2358.825957613841 1.0 33.0 4.0 1.0 1034.1791561981606 1.0 2.0 3.0 1.0 1063.7876971394164 1.0 7.0 4.0 2.0 2734.910

In [0]:
#Let's save this dataframe as a gold and final table: this is the table pre-processed for ML problem
df.write.format("delta").mode("overwrite").saveAsTable("gold_table")

In [0]:
# Split the data into training and test sets
(train_df, test_df) = df.randomSplit([0.8, 0.2])
    
    
# Print the number of records
print(f'There are {train_df.cache().count()} records in the training dataset.')
print(f'There are {test_df.cache().count()} records in the testing dataset.')

There are 124 records in the training dataset.
There are 31 records in the testing dataset.


In [0]:
#Import the necessary libraries
import pyspark
from pyspark.ml import Pipeline
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.regression import LinearRegression, DecisionTreeRegressor, RandomForestRegressor
import mlflow
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.feature import StandardScaler

#We will have one experiment where we will group all of the runs, each containing a different estimator/regressor (linear regression, rf and dt)
experiment_name = "/Users/mviol.exstudents2022@student.ie.edu/our_models_final_project"
experiment_id = mlflow.create_experiment(experiment_name)

with mlflow.start_run(run_name="model_LR", experiment_id=experiment_id) as run:
    vector_assembler = VectorAssembler(inputCols=['review_count', 'categories_index', 'price', 'distance'], outputCol='features')
    # Standardize the columns using StandardScaler
    scaler = StandardScaler(inputCol='features', outputCol='scaled_features', withStd=True, withMean=False)

    

    # Print the number of records
    print(f'There are {train_df.cache().count()} records in the training dataset.')
    print(f'There are {test_df.cache().count()} records in the testing dataset.')


    # Define the regressors
    lr = LinearRegression(featuresCol='scaled_features', labelCol='rating', predictionCol="prediction_lr")
    

    # Define the pipeline
    pipeline = Pipeline(stages=[vector_assembler, scaler, lr])

    # Train the model
    print(train_df)
    model = pipeline.fit(train_df)
    
    
    # Log parameters
    mlflow.log_param("target_variable", "rating")

    # Log the model for this run
    mlflow.spark.log_model(model, "SparkML-linear-regression")

    # Evaluate the model on the test set
    lr_predictions = model.transform(test_df)
    
    
    # Save the prediction as csv
    lr_predictions.toPandas().to_csv('lr_predictions.csv', index=False)
   
    
    # Log the saved prediction as artifact
    mlflow.log_artifact("lr_predictions.csv")

    
    
    # Evaluate predictions for linear regression
    regressionEvaluator_lr = RegressionEvaluator(predictionCol="prediction_lr", labelCol="rating")
    rmse_lr = regressionEvaluator_lr.setMetricName("rmse").evaluate(lr_predictions)
    r2_lr = regressionEvaluator_lr.setMetricName("r2").evaluate(lr_predictions)
    mse_lr = regressionEvaluator_lr.setMetricName("mse").evaluate(lr_predictions)
    mae_lr = regressionEvaluator_lr.setMetricName("mae").evaluate(lr_predictions)
    
    # Log metrics
    mlflow.log_metric("rmse", rmse_lr)  
    mlflow.log_metric("mae", mae_lr)    
    


There are 124 records in the training dataset.
There are 31 records in the testing dataset.
DataFrame[review_count: double, rating: double, price: double, distance: double, categories_index: double]
2022/12/19 19:21:57 WARNING mlflow.utils.environment: Encountered an unexpected error while inferring pip requirements (model URI: /tmp/tmpmw1khryu, flavor: spark), fall back to return ['pyspark==3.3.0']. Set logging level to DEBUG to see the full traceback.
/databricks/spark/python/pyspark/sql/pandas/conversion.py:119: UserWarning: toPandas attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Unable to convert the field features. If this column is not necessary, you may consider dropping it or converting to primitive type before the conversion.
Direct cause: Unsupported type in conversion to Arrow: VectorUDT()
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.

In [0]:
#For decision trees and random forest scaling is not required, but it doesn't hurt so we didn't remove it.

with mlflow.start_run(run_name="model_dt",  experiment_id=experiment_id) as run:
    vector_assembler = VectorAssembler(inputCols=['review_count', 'categories_index', 'price', 'distance'], outputCol='features')
    # Standardize the columns using StandardScaler
    scaler = StandardScaler(inputCol='features', outputCol='scaled_features', withStd=True, withMean=False)

    

    # Print the number of records
    print(f'There are {train_df.cache().count()} records in the training dataset.')
    print(f'There are {test_df.cache().count()} records in the testing dataset.')


    # Define the regressors
    dt = DecisionTreeRegressor(featuresCol='scaled_features', labelCol='rating', predictionCol="prediction_dt")
    
    
    # Define the pipeline
    pipeline = Pipeline(stages=[vector_assembler, scaler, dt])

    # Train the model
    print(train_df)
    model = pipeline.fit(train_df)
    
    
    # Log parameters
    mlflow.log_param("target_variable", "rating")

    # Log the model for this run
    mlflow.spark.log_model(model, "SparkML-decision-tree")

    # Evaluate the model on the test set
    dt_predictions = model.transform(test_df)
    
    # Save the prediction as csv
    dt_predictions.toPandas().to_csv('dt_predictions.csv', index=False)
    
    # Log the saved prediction as artifact
    mlflow.log_artifact("dt_predictions.csv")
    

    # Evaluate predictions for decision tree
    regressionEvaluator_dt = RegressionEvaluator(predictionCol="prediction_dt", labelCol="rating")
    rmse_dt = regressionEvaluator_dt.setMetricName("rmse").evaluate(dt_predictions)
    r2_dt = regressionEvaluator_dt.setMetricName("r2").evaluate(dt_predictions)
    mse_dt = regressionEvaluator_dt.setMetricName("mse").evaluate(dt_predictions)
    mae_dt = regressionEvaluator_dt.setMetricName("mae").evaluate(dt_predictions)
    
    # Log metrics
    mlflow.log_metric("rmse", rmse_dt)  
    mlflow.log_metric("mae", mae_dt)   

    

There are 124 records in the training dataset.
There are 31 records in the testing dataset.
DataFrame[review_count: double, rating: double, price: double, distance: double, categories_index: double]
2022/12/19 19:22:31 WARNING mlflow.utils.environment: Encountered an unexpected error while inferring pip requirements (model URI: /tmp/tmp9mferwsf, flavor: spark), fall back to return ['pyspark==3.3.0']. Set logging level to DEBUG to see the full traceback.
/databricks/spark/python/pyspark/sql/pandas/conversion.py:119: UserWarning: toPandas attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Unable to convert the field features. If this column is not necessary, you may consider dropping it or converting to primitive type before the conversion.
Direct cause: Unsupported type in conversion to Arrow: VectorUDT()
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.

In [0]:
#For decision trees and random forest scaling is not required, but it doesn't hurt so we didn't remove it.

with mlflow.start_run(run_name="model_rf", experiment_id=experiment_id) as run:
    vector_assembler = VectorAssembler(inputCols=['review_count', 'categories_index', 'price', 'distance'], outputCol='features')
    # Standardize the columns using StandardScaler
    scaler = StandardScaler(inputCol='features', outputCol='scaled_features', withStd=True, withMean=False)


    # Print the number of records
    print(f'There are {train_df.cache().count()} records in the training dataset.')
    print(f'There are {test_df.cache().count()} records in the testing dataset.')


    # Define the regressors
    rf = RandomForestRegressor(featuresCol='scaled_features', labelCol='rating', predictionCol="prediction_rf")
    
    
    # Define the pipeline
    pipeline = Pipeline(stages=[vector_assembler, scaler, rf])

    # Train the model
    print(train_df)
    model = pipeline.fit(train_df)
    
    
    # Log parameters
    mlflow.log_param("target_variable", "rating")

    # Log the model for this run
    mlflow.spark.log_model(model, "SparkML-random-forest")

    # Evaluate the model on the test set
    rf_predictions = model.transform(test_df)
    
    # Save the prediction as csv
    rf_predictions.toPandas().to_csv('rf_predictions.csv', index=False)
    
    # Log the saved prediction as artifact
    mlflow.log_artifact("rf_predictions.csv")
    
    

    
    # Evaluate predictions for rf
    regressionEvaluator_rf = RegressionEvaluator(predictionCol="prediction_rf", labelCol="rating")
    rmse_rf = regressionEvaluator_rf.setMetricName("rmse").evaluate(rf_predictions)
    r2_rf = regressionEvaluator_rf.setMetricName("r2").evaluate(rf_predictions)
    mse_rf = regressionEvaluator_rf.setMetricName("mse").evaluate(rf_predictions)
    mae_rf = regressionEvaluator_rf.setMetricName("mae").evaluate(rf_predictions)
    
    # Log metrics
    mlflow.log_metric("rmse", rmse_rf)  
    mlflow.log_metric("mae", mae_rf)   


There are 124 records in the training dataset.
There are 31 records in the testing dataset.
DataFrame[review_count: double, rating: double, price: double, distance: double, categories_index: double]
2022/12/19 19:23:09 WARNING mlflow.utils.environment: Encountered an unexpected error while inferring pip requirements (model URI: /tmp/tmp_b1g0_k7, flavor: spark), fall back to return ['pyspark==3.3.0']. Set logging level to DEBUG to see the full traceback.
/databricks/spark/python/pyspark/sql/pandas/conversion.py:119: UserWarning: toPandas attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:
  Unable to convert the field features. If this column is not necessary, you may consider dropping it or converting to primitive type before the conversion.
Direct cause: Unsupported type in conversion to Arrow: VectorUDT()
Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.

In [0]:
from mlflow.tracking import MlflowClient
MlflowClient().list_experiments()

<command-221565742489074>:2: FutureWarning: ``mlflow.tracking.client.MlflowClient.list_experiments`` is deprecated. This method will be removed in a future release. Use ``search_experiments()`` instead.
  MlflowClient().list_experiments()
Out[67]: [<Experiment: artifact_location='dbfs:/databricks/mlflow-tracking/221565742489086', creation_time=1671477686020, experiment_id='221565742489086', last_update_time=1671477753556, lifecycle_stage='active', name='/Users/mviol.exstudents2022@student.ie.edu/our_models_final_project', tags={'mlflow.experiment.sourceName': '/Users/mviol.exstudents2022@student.ie.edu/our_models_final_project',
  'mlflow.experimentType': 'MLFLOW_EXPERIMENT',
  'mlflow.ownerEmail': 'mviol.exstudents2022@student.ie.edu',
  'mlflow.ownerId': '6801149048339153'}>,
 <Experiment: artifact_location='dbfs:/databricks/mlflow-tracking/221565742489080', creation_time=1671475981474, experiment_id='221565742489080', last_update_time=1671476050887, lifecycle_stage='active', name='/

In [0]:
# Get all runs for a given experiment
experiment_id = run.info.experiment_id
runs_df = mlflow.search_runs(experiment_id)

# Display information
runs_df.T

0 
 1 
 2 
 
 
 
 
 run_id 
 ee85c71d39c648cdb4e92f229ce9483a 
 7058e86692a0431e8d609051e1fd9fe6 
 98a49dd76bea42ef95687a18b3151005 
 
 
 experiment_id 
 221565742489086 
 221565742489086 
 221565742489086 
 
 
 status 
 FINISHED 
 FINISHED 
 FINISHED 
 
 
 artifact_uri 
 dbfs:/databricks/mlflow-tracking/2215657424890... 
 dbfs:/databricks/mlflow-tracking/2215657424890... 
 dbfs:/databricks/mlflow-tracking/2215657424890... 
 
 
 start_time 
 2022-12-19 19:22:33.556000+00:00 
 2022-12-19 19:21:59.225000+00:00 
 2022-12-19 19:21:26.063000+00:00 
 
 
 end_time 
 2022-12-19 19:23:12.332000+00:00 
 2022-12-19 19:22:33.347000+00:00 
 2022-12-19 19:21:59.043000+00:00 
 
 
 metrics.mae 
 0.271531 
 0.321453 
 0.279372 
 
 
 metrics.rmse 
 0.323026 
 0.424287 
 0.329652 
 
 
 params.target_variable 
 rating 
 rating 
 rating 
 
 
 tags.mlflow.databricks.notebookRevisionID 
 1671477792575 
 1671477753508 
 1671477719202 
 
 
 tags.mlflow.databricks.workspaceID 
 3013238818558545 
 3013238818558545 
 3013238818558545 
 
 
 tags.mlflow.source.name 
 /Users/mviol.exstudents2022@student.ie.edu/Fin... 
 /Users/mviol.exstudents2022@student.ie.edu/Fin... 
 /Users/mviol.exstudents2022@student.ie.edu/Fin... 
 
 
 tags.mlflow.databricks.notebookPath 
 /Users/mviol.exstudents2022@student.ie.edu/Fin... 
 /Users/mviol.exstudents2022@student.ie.edu/Fin... 
 /Users/mviol.exstudents2022@student.ie.edu/Fin... 
 
 
 tags.mlflow.databricks.notebookID 
 82594327502686 
 82594327502686 
 82594327502686 
 
 
 tags.sparkDatasourceInfo 
 path=dbfs:/FileStore/tables/df_silver.csv,form... 
 path=dbfs:/FileStore/tables/df_silver.csv,form... 
 path=dbfs:/FileStore/tables/df_silver.csv,form... 
 
 
 tags.mlflow.log-model.history 
 [{"artifact_path":"SparkML-random-forest","fla... 
 [{"artifact_path":"SparkML-decision-tree","fla... 
 [{"artifact_path":"SparkML-linear-regression",... 
 
 
 tags.mlflow.databricks.notebook.commandID 
 5758085053731996640_9203524734153436786_eb5c1f... 
 5758085053731996640_5075229127260517653_3d0e22... 
 5758085053731996640_8130618675562189933_27fe24... 
 
 
 tags.mlflow.source.type 
 NOTEBOOK 
 NOTEBOOK 
 NOTEBOOK 
 
 
 tags.mlflow.databricks.webappURL 
 https://community.cloud.databricks.com 
 https://community.cloud.databricks.com 
 https://community.cloud.databricks.com 
 
 
 tags.mlflow.databricks.cluster.libraries 
 {"installable":[],"redacted":[]} 
 {"installable":[],"redacted":[]} 
 {"installable":[],"redacted":[]} 
 
 
 tags.mlflow.user 
 mviol.exstudents2022@student.ie.edu 
 mviol.exstudents2022@student.ie.edu 
 mviol.exstudents2022@student.ie.edu 
 
 
 tags.mlflow.databricks.workspaceURL 
 https://community.cloud.databricks.com 
 https://community.cloud.databricks.com 
 https://community.cloud.databricks.com 
 
 
 tags.mlflow.runName 
 model_rf 
 model_dt 
 model_LR 
 
 
 tags.mlflow.databricks.cluster.info 
 {"cluster_name":"final_projectt","spark_versio... 
 {"cluster_name":"final_projectt","spark_versio... 
 {"cluster_name":"final_projectt","spark_versio... 
 
 
 tags.mlflow.databricks.cluster.id 
 1219-174551-5q7bnz6o 
 1219-174551-5q7bnz6o 
 1219-174551-5q7bnz6o

We see that that random forest has the lowest rmse and mae, so it has the best performance between all the tested models. However, when we performed other runs we found that linear regression was a good competitor as well (and we see here it is pretty close)

In [0]:
# Get the the latest run
runs = MlflowClient().search_runs(experiment_id, order_by=["metrics.mae asc"], max_results=1)
 
# Get the metrics from the latest run
runs[0].data.metrics

Out[69]: {'mae': 0.27153090764958965, 'rmse': 0.3230257782717405}

Here we could order by time to get latest results, but to make it a bit different we ordered by the model having the best performance (in this case mae, but could also use rmse). Let's interpret these results and what they mean for us. This means that on this occasion we are making an error in rating prediction by approximately 0.27 (if we consider mae) to 0.32 (if we consider rmse which penalizes big errors more). And since our scale of ratings is from 0 to 5, we believe that this is very good since in practice if a restaurant is rated 4.27 or 4.0, it wouldn't matter that much for the user who is going to the restaurant. We conclude that random forest is the best model on this occasion and that a variation of 0.27 on average is very good considering the scale of the ratings.